### PR TITLE
plugins/nvim-cmp: refactor + tests

### PR DESCRIPTION
**Pull request by GaetanLepage**
Refactoring of the `nvim-cmp` plugin.
It now doesn't suffer from the [new `toLuaObject` behavior](https://github.com/pta2002/nixvim/pull/213).
I also made sure that all of the options were correct regarding upstream.

### CODE REVIEW

**Review by traxys — COMMENTED** [review]
Could we have some tests so that we don't break things going forwards ?

**Comment by GaetanLepage** [comment]
> Could we have some tests so that we don't break things going forwards ?

Yes, I plan to write several tests in the following days.

**Comment by GaetanLepage** [comment]
Wow, I found so many errors when writing the test !

File: plugins/completion/nvim-cmp/default.nix [hunk 1]
```diff
@@ -115,103 +79,158 @@ in {
...
-          keyword_length = mkOption {
-            default = null;
-            type = types.nullOr types.int;
-          };
+    mappingPresets = mkOption {
+      default = [];
+      type = types.listOf (types.enum [
+        "insert"
+        "cmdline"
+        # Not sure if there are more or if this should just be str
```
Comment (traxys) [inline plugins/completion/nvim-cmp/default.nix:87]: There are no other presets (lua/cmp/config/mapping.lua)

File: plugins/completion/nvim-cmp/default.nix [hunk 2]
```diff
@@ -115,103 +79,158 @@ in {
...
+          ```
+          function(args)
+            vim.fn["vsnip#anonymous"](args.body) -- For `vsnip` users.
+            -- require('luasnip').lsp_expand(args.body) -- For `luasnip` users.
+            -- require('snippy').expand_snippet(args.body) -- For `snippy` users.
+            -- vim.fn["UltiSnips#Anon"](args.body) -- For `ultisnips` users.
+          end
+          ```
+        '';
+    };
```
Comment (traxys) [inline plugins/completion/nvim-cmp/default.nix:118]: I feel the problem with this one is  that we don't really know which engine is selected so we can't install it. Maybe in a future commit we could be a bit more intelligent with an enum for the diffrent possible engines, and still leave an escape hatch to plug your own ?

File: plugins/completion/nvim-cmp/default.nix (hunk 1 above)
Comment (GaetanLepage) [reply plugins/completion/nvim-cmp/default.nix:87]: Ok thanks ! fixed :)

File: plugins/completion/nvim-cmp/default.nix (hunk 2 above)
Comment (GaetanLepage) [reply plugins/completion/nvim-cmp/default.nix:118]: Ok, I'll do it in another PR.

File: plugins/completion/nvim-cmp/default.nix [hunk 3]
```diff
@@ -115,103 +79,157 @@ in {
...
-          completeopt = mkOption {
-            default = null;
-            type = types.nullOr types.str;
-          };
-        };
-      }));
+      keywordPattern =
+        helpers.defaultNullOpts.mkStr
+        "[[\%(-\?\d\+\%(\.\d\+\)\?\|\h\w*\%(-\w*\)*\)]]"
+        "The default keyword pattern.";
```
Comment (traxys) [inline plugins/completion/nvim-cmp/default.nix:127]: Same tretament as todo-comments, mkRaw with [[ ]] appendended, and  a note in the documentation

**Review by traxys — CHANGES_REQUESTED** [review]

File: plugins/completion/nvim-cmp/default.nix (hunk 3 above)
Comment (GaetanLepage) [reply plugins/completion/nvim-cmp/default.nix]: Done :)

**Review by traxys — APPROVED** [review]
LGTM

**Pull request merged**